# Cancer pipeline

### Resources and knowledgebase

1. Mutation Identification Pipeline (MIP):
	* http://mip-api.readthedocs.io/en/latest/index.html
	* https://github.com/Clinical-Genomics/MIP 
2. ClinSeq autoseq:
	* https://github.com/ClinSeq/autoseq 
3. MSK-Impact pipeline: https://www.mskcc.org/msk-impact
4. TCGA: https://cancergenome.nih.gov/
5. COSMIC: http://cancer.sanger.ac.uk/cosmic
6. CAW: https://github.com/SciLifeLab/CAW
7. Probablistic2020: https://github.com/KarchinLab/probabilistic2020 and it's sister, 2020plus: https://github.com/KarchinLab/2020plus
8. dbSNP
9. ClinVar
10. ExAC
11. GTEx
12. OMIM
13. Drug resistance database
14. Mutational pattern
15. DGVa
16. https://github.com/genome/cancer-genomics-workflow/wiki

---

## Papers

#### Methods and pipelines:
1. **MSK-IMPACT papers:**
	* Cheng, D. T., Mitchell, T. N., Zehir, A., Shah, R. H., Benayed, R., Syed, A., … Berger, M. F. (2015). Memorial sloan kettering-integrated mutation profiling of actionable cancer targets (MSK-IMPACT): A hybridization capture-based next-generation sequencing clinical assay for solid tumor molecular oncology. Journal of Molecular Diagnostics, 17(3), 251–264. https://doi.org/10.1016/j.jmoldx.2014.12.006
	* Cheng, D. T., Prasad, M., Chekaluk, Y., Benayed, R., Sadowska, J., Zehir, A., … Zhang, L. (2017). Comprehensive detection of germline variants by MSK-IMPACT, a clinical diagnostic platform for solid tumor molecular oncology and concurrent cancer predisposition testing. BMC Medical Genomics, 10(1), 33. https://doi.org/10.1186/s12920-017-0271-4
2. **Application of MSK-IMPACT:** Zehir, A., Benayed, R., Shah, R. H., Syed, A., Middha, S., Kim, H. R., … Berger, M. F. (2017). Mutational landscape of metastatic cancer revealed from prospective clinical sequencing of 10,000 patients. Nature Medicine, 23(6), 703–713. https://doi.org/10.1038/nm.4333
3. **VarDict:** Lai, Z., Markovets, A., Ahdesmaki, M., Chapman, B., Hofmann, O., Mcewen, R., … Dry, J. R. (2016). VarDict: A novel and versatile variant caller for next-generation sequencing in cancer research. Nucleic Acids Research, 44(11), 1–11. https://doi.org/10.1093/nar/gkw227
4. **Review on bioinformatic pipelins:** Leipzig, J. (2017). A review of bioinformatic pipeline frameworks. Briefings in Bioinformatics, 18(3), 530–536. https://doi.org/10.1093/bib/bbw020
5. **Mutational signature reviews:**
	* Helleday, T., Eshtad, S., & Nik-Zainal, S. (2014). Mechanisms underlying mutational signatures in human cancers. Nature Reviews Genetics, 15(9), 585–598. https://doi.org/10.1038/nrg3729
	* Alexandrov, L. B., & Stratton, M. R. (2014). Mutational signatures: The patterns of somatic mutations hidden in cancer genomes. Current Opinion in Genetics and Development, 24(1), 52–60. https://doi.org/10.1016/j.gde.2013.11.01
6. 

#### Reviews (including methodological benchmarking):

#### Case studies:

---

## Tools:

1. Teaser: NGS readmapping benchmarking.
	* http://teaser.cibiv.univie.ac.at/
	* https://github.com/Cibiv/Teaser
2. FastQC: Quality control tool. https://www.bioinformatics.babraham.ac.uk/projects/fastqc/
3. Cutadapt: Adapter removal tool. https://cutadapt.readthedocs.io/en/stable/
4. Trim Galore!: FastQC and Cutadapt wrapper. https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/
5. Picardtools: BAM/SAM/VCF/CRAM manipulator. http://broadinstitute.github.io/picard/
	* MarkDuplicate: Mark duplicate reads and potentially remove them
	* LiftoverVcf: liftover VCF between builds
	* CollectHsMetric: Collects hybrid-selection (HS) metrics for a SAM or BAM file
	* CollectAlignmentSummaryMetrics: Produces a summary of alignment metrics from a SAM or BAM file
	* CollectGcBiasMetrics: Collect metrics regarding GC bias
	* CollectWgsMetrics: Collect metrics about coverage and performance of whole genome sequencing (WGS) experiments
6. GATK: A variant discovery tool: https://software.broadinstitute.org/gatk/
	* BaseRecalibrator: Detect systematic error in base quality score
	* Somatic Indel Realigner: Local Realignment around Indels
	* ContEst: Estimate cross sample contamination
	* DepthOfCoverage: Assess sequence coverage by sample, read group, or libraries
	* DuplicateReadFilter: remove duplicated from flag set by MarkDuplicates
7. Samtools: Reading/writing/editing/indexing/viewing SAM/BAM/CRAM format http://www.htslib.org/
8. Sambamba: Tools for working with SAM/BAM/CRAM data http://lomereiter.github.io/sambamba/
9. bcftools: Reading/writing BCF2/VCF/gVCF files and calling/filtering/summarising SNP and short indel sequence variants http://www.htslib.org/doc/bcftools.html
10. vcftools: VCFtools is a program package designed for working with VCF files, such as those generated by the 1000 Genomes Project. https://vcftools.github.io/index.html
11. Delly2: An integrated structural variant prediction method that can discover, genotype and visualize deletions, tandem duplications, inversions and translocations https://github.com/dellytools/delly
12. PLINK: PLINK: Whole genome data analysis toolset https://www.cog-genomics.org/plink2
13. freebayes: a haplotype-based variant detector. https://github.com/ekg/freebayes
14. ASCAT: Allele-Specific Copy Number Analysis of Tumors https://github.com/Crick-CancerGenomics/ascat
15. MutationalPatterns: R package for extracting and visualizing mutational patterns in base substitution catalogues https://github.com/UMCUGenetics/MutationalPatterns
16. desconstructSigs: identification of mutational signatures within a single tumor sample https://github.com/raerose01/deconstructSigs
17. treeOmics: Decrypting somatic mutation patterns to reveal the evolution of cancer
 https://github.com/johannesreiter/treeomics
18. controlFreeC: Copy number and allelic content caller http://boevalab.com/FREEC/
19. MuTect2: Call somatic SNPs and indels via local re-assembly of haplotypes https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_gatk_tools_walkers_cancer_m2_MuTect2.php
20. Annovar: annotation of detected genetic variation http://annovar.openbioinformatics.org/en/latest/
21. Strelka: Small variant caller https://github.com/Illumina/strelka
22. Manta: Structural variant caller https://github.com/Illumina/manta
23. PurBayes: estimate tumor purity and clonality
24. VarDict: variant caller for both single and paired sample variant calling from BAM files https://github.com/AstraZeneca-NGS/VarDict
25. SNPeff: Genomic variant annotations and functional effect prediction toolbox. http://snpeff.sourceforge.net/
26. IGV: visualization tool for interactive exploration http://software.broadinstitute.org/software/igv/
27. SVDetect: a tool to detect genomic structural variations http://svdetect.sourceforge.net/Site/Home.html
28. Genome STRiP: A suite of tools for discovering and genotyping structural variations using sequencing data http://software.broadinstitute.org/software/genomestrip/
29. BreakDancer: SV detection from paired end reads mapping https://github.com/genome/breakdancer
30. pIndel: Detect breakpoints of large deletions, medium sized insertions, inversions, and tandem duplications https://github.com/genome/pindel
31. VarScan:  Variant calling and somatic mutation/CNV detection https://github.com/dkoboldt/varscan
32. VEP: Variant Effect Predictor https://www.ensembl.org/info/docs/tools/vep/index.html
